demos/python_demos/common/performance_metrics.py

```diff
@@ -0,0 +1,97 @@
+"""
+ Copyright (C) 2020 Intel Corporation
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+from time import perf_counter
+import cv2
+from helpers import put_highlighted_text
+
+
+class Statistic:
+    def __init__(self):
+        self.latency = 0.0
+        self.period = 0.0
+        self.frame_count = 0
+
+
+class PerformanceMetrics:
+    def __init__(self, time_window=1.0):    # defines the length of the timespan used for 'current fps' calculation,
+        self.time_window_size = time_window # set to '1 second' by default
+        self.last_moving_statistic = Statistic()
+        self.current_moving_statistic = Statistic()
+        self.total_statistic = Statistic()
+        self.last_update_time = None
+    
+    def update(self, last_request_start_time, frame, position=(15, 30),
+               font_scale=0.75, color=(200, 10, 10), thickness=2):
+        current_time = perf_counter()
+        if self.last_update_time is None:
+            self.last_update_time = current_time
+
+        last_request_latency = current_time - last_request_start_time
+        if self.total_statistic.frame_count != 0:
+            self.current_moving_statistic.latency += last_request_latency
+            self.current_moving_statistic.period = current_time - self.last_update_time
+            self.current_moving_statistic.frame_count += 1
+        else:
+            # for 1st frame
+            self.last_moving_statistic.latency = last_request_latency
+            self.last_moving_statistic.frame_count = 1
+            self.total_statistic.latency = last_request_latency
+            self.total_statistic.frame_count = 1
```

Comment:
It's for cases when the input consists of only one image.

>And it seems that this code sets frame_count to 1 while period is still 0 - doesn't that result in infinite FPS?

Yes, it results in infinite FPS, but it's handled in `get_last` and `get_total`. They return `None` in this case. So, the incorrect value is not shown.


based on Wovchena's comment (https://github.com/opencv/open_model_zoo/pull/1222#pullrequestreview-430229272)
>The special case is latency and fps for the first frame. Handling it is important because the input may consist of one image and some demos allow setting 0 as an argument for waitKey(). My opinion is that moving latency should be initialized for the first time after the first frame is encountered but not after time_window_size passes. There are two options for fps: report 1/latency or report nothing. I like the second option because in that case we don't mix approaches used to compute fps (1/latency for the first and frame frequency for the rest).

Reply:
I guess I understand the motivation, but won't this skew the results in the non-single-image case? E.g., after the second frame is processed, the period will only cover the processing of the second frame, leading to the FPS being reported as being higher than it actually is. This error will decrease as more frames are processed, but this logic still feels wrong.

Maybe it would be better to record the start time in the constructor after all? (Or to have a dedicated `start()` method.) That way, no special first-frame logic would be required.

Also, regarding the demos with `waitKey(0)`... Assuming we do a similar thing to #1229 for the Python demos, won't demos no longer need to do that? I was under the impression that for images, we'll just tell the users to use `-loop_input` and have the demos process the same image over and over. That way, we'll likely get more accurate performance metrics, as well as remove the need for any special logic. @Wovchena Was that your plan or am I misunderstanding?

Reply:
> I guess I understand the motivation, but won't this skew the results in the non-single-image case? E.g., after the second frame is processed, the period will only cover the processing of the second frame, leading to the FPS being reported as being higher than it actually is. This error will decrease as more frames are processed, but this logic still feels wrong.

I think the question is where to begin counting. I assume you expect timer to be started at pipeline start for FPS. But FPS is frame frequency, and I believe we should start counting when the first frame is shown. It's similar to reporting FPS for the first frame - it's undefined because there is no previous frame.

> Also, regarding the demos with waitKey(0)... Assuming we do a similar thing to #1229 for the Python demos, won't demos no longer need to do that? I was under the impression that for images, we'll just tell the users to use -loop_input and have the demos process the same image over and over. That way, we'll likely get more accurate performance metrics, as well as remove the need for any special logic. @Wovchena Was that your plan or am I misunderstanding?

I'm not sure if we will completely remove the option to change `waitKey()` argument for now, but that makes seance. On the other hand we still need the implemented approach because the demo won't report metrics for the first `time_window` seconds otherwise.

Reply:
> I assume you expect timer to be started at pipeline start for FPS. But FPS is frame frequency, and I believe we should start counting when the first frame is shown.

Yeah, that's fair enough.

I think the crux of the issue here is that the two stats we measure are properties of different things: the latency is the property of a frame, while the period is the interval _between_ frames. And so once N frames have been processed, we will naturally have N latencies, but only N-1 periods. Therefore, it is incorrect to divide by `frame_count` when calculating total averages, because no matter how you count frames, one of the averages will be wrong.

I see two reasonable ways to resolve this:

1. Just don't collect the latency for the first frame. Then we'll always have the same number of latencies and periods. This does mean that we won't be able to report the latency for the first `time_window` seconds. I don't think that's a huge problem, because the latency of the first frame is likely to be non-representative anyway (since the pipeline won't have stabilized yet). This approach has the advantage of code being simpler, and it makes sure that the stats the user sees are always calculated the same way (i.e., by averaging).

2. Instead of the block this comment thread is attached to, save the first frame's latency to a separate variable (e.g., self.first_frame_latency). In `get_total`, add `self.first_frame_latency` to the numerator of the latency, and 1 to the denominator. In `get_last`, check if `self.total_statistic.frame_count == 0`, and if so, return `(self.first_frame_latency, None)`. This is pretty ugly, but it avoids the error in the FPS calculation, and faithfully represents the fact that we calculate the latency differently for the first `time_window` seconds.

Reply:
I didn't notice that problem.

I prefer the first way of resolving.